point.go

```diff
@@ -0,0 +1,306 @@
+	} else {
+		dst[0] = byte(tmp.y.n[0] & 1)
+	}
+
+	putB32From5x52(dst[1:PointSizeMarshalled], &tmp.x)
+}
+
+// SetBytes sets the field element to be equal to the given byte slice. It will
+// return an error if the data does not represent a valid curve point.
+//
+// Panics: If the byte slice has length less than 33, this function will panic.
+func (p *Point) SetBytes(bs []byte) error {
+	if len(bs) < PointSizeMarshalled {
+		panic(fmt.Sprintf("invalid slice length: length needs to be at least 33, got %v", len(bs)))
+	}
+
+	if bs[0] == 0xFF {
+		p.inner.infinity = 1
+		return nil
+	}
+
+	var tmp C.secp256k1_ge
+
+	set5x52FromB32(bs[1:PointSizeMarshalled], &tmp.x)
+	if C.secp256k1_ge_set_xo_var(&tmp, &tmp.x, C.int(bs[0])&1) == 0 {
+		// The x coordinate does not correspond to a valid curve point.
+		C.secp256k1_gej_clear(&p.inner)
+		return errors.New("invalid curve point data")
+	}
+
+	// After reconstructing the y coordinate, it is not guaranteed to be
+	// normalized, so we do that manually.
+	C.secp256k1_fe_normalize_var(&tmp.y)
+
+	C.secp256k1_gej_set_ge(&p.inner, &tmp)
+
+	return nil
+}
+
+// SizeHint implements the surge.SizeHinter interface.
+func (p Point) SizeHint() int { return PointSizeMarshalled }
+
+// Marshal implements the surge.Marshaler interface.
+func (p Point) Marshal(buf []byte, rem int) ([]byte, int, error) {
+	if len(buf) < PointSizeMarshalled || rem < PointSizeMarshalled {
+		return buf, rem, surge.ErrUnexpectedEndOfBuffer
+	}
+
+	p.PutBytes(buf[:PointSizeMarshalled])
+
+	return buf[PointSizeMarshalled:], rem - PointSizeMarshalled, nil
+}
+
+// Unmarshal implements the surge.Unmarshaler interface.
+func (p *Point) Unmarshal(buf []byte, rem int) ([]byte, int, error) {
+	if len(buf) < PointSizeMarshalled || rem < PointSize {
+		return buf, rem, surge.ErrUnexpectedEndOfBuffer
+	}
+
+	err := p.SetBytes(buf[:PointSizeMarshalled])
+	if err != nil {
+		return buf[PointSizeMarshalled:], rem - PointSize, err
+	}
+
+	return buf[PointSizeMarshalled:], rem - PointSize, nil
+}
+
+// IsInfinity returns true if the point represents the point at infinity, and
+// false otherwise.
+func (p *Point) IsInfinity() bool {
+	return p.inner.infinity != 0
+}
+
+// IsOnCurve returns true if the point is on the elliptic curve, and false
+// otherwise. The point at infinity will return false.
+func (p *Point) IsOnCurve() bool {
+	var tmp C.secp256k1_ge
+	C.secp256k1_ge_set_gej(&tmp, &p.inner)
+	return C.secp256k1_ge_is_valid_var(&tmp) != 0
+}
+
+// Eq returns true if the two curve points are equal, and false otherwise.
+func (p *Point) Eq(other *Point) bool {
+	if p.IsInfinity() != other.IsInfinity() {
+		return false
+	}
+
+	if p.IsInfinity() {
+		return true
+	}
+
+	// Scale p so that the z fields are equal. Once the z fields are equal, the
+	// points will be equal if and only if the x and y fields are equal.
+	//
+	// NOTE: This modifies the representation of p, but it will still represent
+	// the same point on the elliptic curve.
```

Comment:
This note should probably go in the documenting comment so that it shows up in godoc and in IDEs.

Reply:
Fixed in e5a665a.